noir-projects/noir-protocol-circuits/crates/rollup-lib/src/root.nr

```diff
@@ -6,15 +6,16 @@ use root_rollup_inputs::RootRollupInputs;
 use root_rollup_public_inputs::RootRollupPublicInputs;
 
 // TODO: Move all the following code to different files
-use dep::types::{constants::{NUM_FIELDS_PER_SHA256, NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP}, utils::uint256::U256};
+use dep::types::{constants::{NUM_FIELDS_PER_SHA256, NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP}, utils::uint256::U256, hash::sha256_to_field};
 
 // See `test_message_input_flattened_length` on keeping this in sync,
 // why its here and how this constant was computed.
 global NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP_NUM_BYTES: u64 = 512;
 
 // Computes the messages hash from the leaves array
 //
-// Returns the hash split into two field elements
+// TODO(Miranda): should this be merged with a generic version of accumulate_sha256? (The logic is largely the same)
```

Comment:
Sounds good!

Reply:
I had a go at this, but found the gas cost would increase quite a lot in replicating for calculating the `inHash`. Basically, `l1toL2Msgs` are stored in bytes32 in solidity so replicating the `accumulate` nr hash requires truncating each one manually over a loop, then concatenating and hashing.
Unfortunately, to avoid casting issues, I've already changed other hashes (like the txEffectHash) to be bytes31 so this creates a discrepancy. Perhaps in another PR I have two choices:

- force all sha hashes that are touched in nr to be bytes31 (requires a lot of changes in ts/sol and possibly higher gas)
- revert the hashes that I have cast to bytes31, and prepend/drop a byte where required in solidity (probably messier code)

I think both are valid choices, so would appreciate a pointer!

Reply:
We have decided to go for 32 bytes everywhere, however we would require two generics in `accumulate_sha256` - one for the number of input fields and one for 32 * that number. Unfortunately Noir doesn't realise 32*N can be known at compile time, so I would need another generic or some constant, which wouldn't make sense.
I did notice though that this function (`compute_messages_hash`) now appears to be unused, so perhaps this is not a problem anymore?